### DISCUSSION TITLE
Request Model Resolved at Runtime -- currently uses type hints

### DISCUSSION THREAD

**Discussion by farzadsbakhtiar** [fastapi/fastapi#5965] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class ModelReadHandler:
  def __init__(self, base_url: str, response_model_cls: type, request_model_cls: type):
    self.router = APIRouter()
    self.response_model = response_model_cls
    self.request_model = request_model_cls
    self.add_routes()
  def add_routes():
    self.router.add_api_route(f'/{self.base_url}/', self.read, methods=['GET'],
                               response_model=self.response_model)
  def read(self, 
           input_model: self.request_model_cls # ** actually can't do this but looking the resulting behaviour
           session: Depends(get_session),
           ):
    query = input_model.get_read_query()
    return session.exec(query).all()

# -- the input_model would extend this abstract class
class InputModel(pydantic.BaseModel):
  def get_read_query(self):
    raise NotImplementedError()

# class example for InputModel
class UserInputModel(InputModel):
  username: int
  def get_read_query(self):
    return select(UserModel).where(UserModel.username == self.username)

# and we instantiate the user handler like this
user_handler = ModelReadHandler(base_url='user',
                                request_model_cls=UserInputModel,
                                response_model_cls=UserModel)

app = FastAPI()
app.include_router(user_handler.router)
```

### Description
We can use `response_model` instead of the **function return type hint** to define the format of the response. I'm looking for the same thing for the **request**, to replace the **function arg type hint**.  Here is some context and description of the code above:  
I'm writing a class-based generic CRUD router/handler to generate a Swagger UI and CRUD endpoints for new models without writing new functions/classes. Above code is an example for reading a user model and filtering it base on it's username.  
Now it doesn't work, because there is no way to set the type of the `input_model` argument in `ModelReadhandler.read()` during runtime. One workaround would be to create a new handler for each new model extending this base handler and supplying an e.g. `.read()` method with the proper type hint. I am trying to avoid this.  
As @caeser1996 notes below, another option is to use the `Request` object directly, but that wouldn't add the endpoint parameters to the Swagger UI.

### Operating System

macOS

### FastAPI Version

0.88.0

### Python Version

3.10

**Comment by caeser1996** [comment]


**Reply by farzadsbakhtiar** [reply]
Hi Caeser,
Thanks for the response.
That works but wouldn't generate a swagger UI.

**Reply by caeser1996** [reply]
Are all your versions up to date(including fast API and python)?

**Reply by farzadsbakhtiar** [reply]
By that, I mean it wouldn't include the parameters of the input model for the endpoint in Swagger UI.

**Comment by baskervilski** [comment, marked as answer]
I don't know if it serves your use case, but I did something like this:

```python

def make_router(request_model, response_model, my_function):
    router = APIRouter()
    
    @router.post("/my_function", response_model=response_model):
    async def api_for_my_function(request: request_model):
        result_dict= my_function(request.some_attribute)
        return response_model(**result_dict)
    
    return router
```

And it gives me the exact request model in the `docs`.
